File: backend-java/src/main/java/com/example/modules/catalog/internal/CatalogService.java
// Business logic — hidden from the outside world.
package com.example.modules.catalog.internal;

import com.example.modules.catalog.infrastructure.SeedData;
import com.example.modules.catalog.infrastructure.SeedData.Product;

import java.util.List;
import java.util.Optional;

public class CatalogService {
    private final List<Product> products = SeedData.seed();

    public List<Product> list(String category) {
        return products.stream()
            .filter(p -> category == null || category.isBlank() || p.category().equals(category))
            .toList();
    }

    public Optional<Product> get(String sku) {
        return products.stream().filter(p -> p.sku().equals(sku)).findFirst();
    }

    public List<Product> search(String query) {
        String q = query.toLowerCase();
        return products.stream()
            .filter(p -> p.name().toLowerCase().contains(q) || p.description().toLowerCase().contains(q))
            .toList();
    }
}

File: backend-csharp/modules/catalog/Internal/CatalogService.cs
// Business logic — hidden from the outside world.
using BackendCsharp.Modules.Catalog.Infrastructure;
namespace BackendCsharp.Modules.Catalog.Internal;

public class CatalogService
{
    private readonly List<Product> _products = SeedData.Seed();

    public IEnumerable<Product> List(string? category) =>
        _products.Where(p => string.IsNullOrEmpty(category) || p.Category == category);

    public Product? Get(string sku) => _products.FirstOrDefault(p => p.Sku == sku);

    public IEnumerable<Product> Search(string query)
    {
        var q = query.ToLowerInvariant();
        return _products.Where(p => p.Name.ToLowerInvariant().Contains(q) || p.Description.ToLowerInvariant().Contains(q));
    }
}
